script/testing/util/db_server.py

```diff
@@ -227,21 +228,220 @@ def get_build_path(build_type):
+    """
+    def wrap_outer(f):
+        def wrap_inner(target, meta):
+            return f(target, meta) if any(isinstance(target, ty) for ty in target_types) else target
+        return wrap_inner
+    return wrap_outer
+
+# -----------------------------------------------------------------------------
+# Attribute Preprocessors
+#
+# The signature for attribute preprocessors must be:
+#   def preprocessor(attr: str, meta: Dict) -> str:
+#       ...
+
+
+# -----------------------------------------------------------------------------
+# Value Preprocessors
+#
+# The signature for value preprocessors must be:
+#   def preprocessor(value: str, meta: Dict) -> str:
+#       ...
+
+@applies_to(bool)
+def lower_booleans(value, meta):
+    """
+    Lower boolean string values to the format expected by the DBMS server.
+    
+    e.g. 
+        `True` -> `true`
+        `False` -> `false`
+    
+    Arguments
+    ---------
+    value : str 
+        The DBMS server argument value
+    meta : Dict
+        Dictionary of meta-information available to all preprocessors
+
+    Returns
+    -------
+    The preprocessed server argument value
+    """
+    return str(value).lower()
 
-    return ' '.join(server_args_arr)
+@applies_to(str)
+def resolve_relative_paths(value, meta):
+    """
+    Resolve relative paths in the DBMS server arguments to their equivalent absolute paths.
+
+    When specifying path arguments to the DBMS, it is often simpler to think in terms
+    of how the path relates to the location of the DBMS server binary. However, because
+    the command used by NoisePageServer to launch the DBMS server instance specifies
+    the absolute path to the binary, relative paths (that necessarily encode the expected
+    location of the binary) will fail to function properly. Expanding these relative paths
+    to their absolute counterparts before passing them to the DBMS addresses this issue.
+
+    The above highlights the fundamental limitation of this current implementation, namely:
+
+        ALL RELATIVE PATHS ARE ASSUMED TO BE RELATIVE TO THE DBMS BINARY DIRECTORY
+
+    All relative paths that are relative to another directory will fail to resolve properly.
+
+    Arguments
+    ---------
+    value : str 
+        The DBMS server argument value
+    meta : Dict
+        Dictionary of meta-information available to all preprocessors
+
+    Returns
+    -------
+    The preprocessed server argument value
+    """
+    is_relative = str.startswith(value, "./") or str.startswith(value, "../")
+    # TODO(Kyle): This doesn't actually do any "resolving", it merely appends
+    # the absolute path to the binary directory to the relative path as it 
+    # presently exists; this works fine, but also means we might pass a needlessly
+    # long string to the DBMS server e.g. bin/././././whatever.txt is possible.
+    # It might be worth it to find some portable way to completely resolve these.
```

Comment:
So I have an issue with this implementation that I have been mulling over, would appreciate an additional perspective. The functionality in this preprocessor is necessary, and I like having it there, but I don't like how 'loose' the constraint on it is - that is, we will attempt to apply this preprocessor to _any_ string that happens to start with `./` or `../` because we assume this means it is a relative path. This is _probably_ a safe assumption, but what I don't like is the fact that it applies to **all strings** and not just strings that are also paths.

I have been trying to come up with a way to express the fact that an argument is a path, rather than just a string. The best I have come up with so far (I think) is just to look at the suffix of the argument and declare that if it ends with `_path` it really is a string that corresponds to a path. We obey this naming scheme consistently for all of the arguments the DBMS currently supports, but of course this introduces a constraint that all future arguments must continue to observe this.

I even though about going so far as to parse the header file where the settings are defined (`settings_defs.h`) and derive type information from that file, but that doesn't really get us anywhere because its not like the string settings defined therein look any different (i.e. use a different macro, have a different type) when they are paths versus when they are not. Of course, this is a change I could make (refactor settings to add support for e.g. `SETTING_path`) but I fear I would be encroaching on over-engineering territory.

Reply:
If you can solve the problem in a single `_hidden_helper` function and <20 lines of code I'd say go for it. Otherwise, generally I do not think people muck around with infrastructure often, and those who do are generally prepared for a little bit of madness given the state of C++ infrastructure, so as long as the blowing up error is obvious then path issues aren't too bad to debug (xyz not found).

If you just want to apply it to paths, maybe you can enforce it at a type level?
https://docs.python.org/3/library/pathlib.html

Reply:
Ended up just implementing the `_path` check. My thinking is, at the worst, this will end up failing to resolve paths that some user may have thought would be resolved, which is better than the alternative. One has to explicitly opt-in to being resolved, rather than the other way around.